### PR TITLE
[Spring 43] Simple Word counter using only available Spring XD components

### PR DESCRIPTION
**Pull request by mbogoevici**
- file reading is handled by a file source (which natively splits content by line)
- word splitting is handled by a splitter
- word counting is done by a field value counter
- visualization through the analytics-dashboard
- variant using a custom counter, for a simpler stream structure

### CODE REVIEW

File: simple-word-count/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+stream create words --definition "file --dir=<directory-with-text-files> --outputType=text/plain | splitter --expression=payload.split(' ')  | log
+```
+
+Then we will tap the original stream and create a new one, for counting the words. We will use a field value counter for accumulating results in the 
+* the `words` stream is tapped after the splitter;
+* we use a transformer to wrap the individual words into Spring XD `Tuple`s, which are one of the expected 
+
+
+```
+stream create wordcount --definition "tap:stream:words.transformer > transform --expression=T(org.springframework.xd.tuple.TupleBuilder).tuple().of('word',payload) | field-value-counter --fieldName=word"
```
Comment (thomasdarimont) [inline simple-word-count/README.md]: I think `tap:stream:words.transformer > transform`should rather be `tap:stream:words.splitter > transform`?

**Comment by thomasdarimont** [comment]
Thanks for this example - it makes a great demo with very little effort :+1: 

I use the following stream definition
```
stream destroy words
stream destroy wordcount
field-value-counter delete wordcount

stream create words --definition "file --dir=/data/text --outputType=text/plain | splitter --expression=payload.split(' ')  | log"

stream create wordcount --definition "tap:stream:words.splitter > transform --expression=T(org.springframework.xd.tuple.TupleBuilder).tuple().of('word',payload.trim()) | field-value-counter --fieldName=word"

stream deploy wordcount
stream deploy words
```
- Note the `payload.trim()` in the before the `field-value-counter`to get rid of some potential whitespace characters.

File: simple-word-count/README.md (hunk 1 above)
Comment (mbogoevici) [inline simple-word-count/README.md]: Oh yes. I've changed that example around a few times - couldn't decide where the transformer should go. It seems a little artificial to have it there, but the plain payload cannot be used for the counter.

**Comment by mbogoevici** [comment]
Thanks @thomasdarimont ! I'm not sure if the `trim()` is necessary here though, as we split by whitespace already.

**Comment by dturanski** [comment]
LGTM; merged

**Comment by markfisher** [comment]
Seeing this: T(org.springframework.xd.tuple.TupleBuilder).tuple().of('word',payload.trim())

...makes me think we should add some custom SpEL functions for ease-of-use for common Tuple-based operations

**Comment by mbogoevici** [comment]
Agreed... There is another way that @dturanski suggested, by constructing a JSON String and converting the result, but eventually this should be easier to use from within SPEL.

**Comment by markfisher** [comment]
All we need is another <int:spel-function/> in parent-context.xml. Seems like low-hanging fruit to add some for Tuples. cc @markpollack @sabbyanandan

**Comment by ericbottard** [comment]
Agreed. Also, creating a tuple from another tuple (for transformation) is
too complicated at the moment

**Comment by dturanski** [comment]
Good point re. creating a tuple from another.  Currently the semantics of
TupleBuilder present a challenge. e.g.:

TupleBuilder.fromTuple(t).of('foo',bar') ?

**Comment by thomasdarimont** [comment]
Would be great to have a section in the wiki (maybe here: https://github.com/spring-projects/spring-xd/wiki/Extending-XD) that shows how to add custom SpEL expressions across modules.

I added the following SpEL functions to the transform module: `modules/processor/transform/config/transform.xml``
```xml
	<spel-function id="parseLong"
					   class="java.lang.Long"
					   method="parseLong(java.lang.String)"/>

	<spel-function id="parseInt"
					   class="java.lang.Integer"
					   method="parseInt(java.lang.String)"/>


	<spel-function id="parseDouble"
					   class="java.lang.Double"
					   method="parseDouble(java.lang.String)"/>
	
	<spel-function id="tuple"
					   class="org.springframework.xd.tuple.TupleBuilder"
					   method="tuple()"/>
```

**Pull request closed without merging**